src/main/java/io/vertx/rabbitmq/RabbitMQClient.java

```diff
@@ -43,6 +46,16 @@ static RabbitMQClient create(Vertx vertx, RabbitMQOptions config) {
     return new RabbitMQClientImpl(vertx, config);
   }
 
+
+  /**
+   * Set a callback to be called whenever a new connection is established.
+   * This callback must be idempotent - it will be called each time a connection is established, which may be multiple times against the same instance.
+   * Callbacks will be added to a list and called in the order they were added, the only way to remove callbacks is to create a new client.
+   * @param connectionEstablishedCallback  callback to be called whenever a new connection is established.
+   */
+  @GenIgnore
+  void addConnectionEstablishedCallback(Handler<Promise<Void>> connectionEstablishedCallback);
```

Comment:
I don't get why the callback will pass a `Promise<Void>` instead of `Void` here, can you elaborate ?

Reply:
The connectionEstablishedCallbacks do two things: they give the user the opportunity to set up exchanges, queues & bindings; and they give the publisher the opportunity to add its confirmListener.
These things are all async and need to be complete before we start trying to publish or consume, so I use the Promise to let the callback tell the client when it should proceed.

Reply:
can you document it ? the documentation does not mention it nor the javadoc, this needs to be documented

Reply:
Added some more to the docs for that method. Any good?

Reply:
you need to say what is the promise for and show how / when it should be completed or failed

Reply:
The latest [commit](https://github.com/Yaytay/vertx-rabbitmq-client/commit/87a16c8942a3b08ce8c85808feb5d3184dcc7680) has this:
  /**
   * Set a callback to be called whenever a new connection is established.
   * This callback must be idempotent - it will be called each time a connection is established, which may be multiple times against the same instance.
   * Callbacks will be added to a list and called in the order they were added, the only way to remove callbacks is to create a new client.
   * 
   * These callbacks should be used to establish any Rabbit MQ server objects that are required - exchanges, queues, bindings, etc.
   * Each callback will receive a Promise<Void> that it must complete in order to pass control to the next callback (or back to the RabbitMQClient).
   * If the callback fails the promise the RabbitMQClient will be unable to make a connection (it will attempt to connect again according to its retry configuration).
   * If the promise is not completed or failed by a callback the RabbitMQClient will not start (it will hang indefinitely).
   * 
   * Other methods on the client may be used in the callback -
   * it is specifically expected that RabbitMQ objects will be declared, but the publish and consume methods must not be used.
   * 
   * The connection established callbacks are particularly important with the RabbitMQPublisher and RabbitMQConsumer when they are used with 
   * servers that may failover to another instance of the server that does not have the same exchanges/queues configured on it.
   * In this situation these callbacks are the only opportunity to create exchanges, queues and bindings before the client will attempt to use them when it 
   * re-establishes connection.
   * If your failover cluster is guaranteed to have the appropriate objects already configured then it is not necessary to use the callbacks.
   * 
   * @param connectionEstablishedCallback  callback to be called whenever a new connection is established.
   */

Do I need to create a new PR or does this include that?